### ISSUE TITLE
Moving too quickly between pages on the popped-out "speaker notes" window resets to non-notes view

### ISSUE THREAD

**Issue by fw-immunant** [google/comprehensive-rust#2004]
To reproduce, tap left and right a couple times while the speaker notes window is focused after popping it out into its own window. It will revert to showing a second copy of the course material, as if the speaker notes window had been closed.

**Comment by mgeisler** [comment]
Thanks for logging this! The speaker notes system is quite rudimentary and it could benefit from someone with better JavaScript skills than me giving it an overhaul.

**Comment by randomPoison** [comment]
I've found that navigating at all in the speaker notes immediately breaks the connection between the speaker notes and the main window. I'm assuming this is the same issue that @fw-immunant ran into here.

**Comment by djmitche** [comment]
I've seen that as well.

**Comment by michael-kerscher** [comment]
I can also reproduce this reliably by clicking into the notes and go to the right. The speaker notes view is now turning into the regular slide view".
Moving left does break the connection to the regular slides but stays in the speaker notes view - it is independent from the regular slides though

Interesting behavior: moving left or right has a different effect

**Comment by michael-kerscher** [comment]
I found the issue with the "next" navigation and created a pull request.

There is an additional issue causing the break:
https://github.com/google/comprehensive-rust/blob/c05f0b6f023bf39b8d7f96bfa63e449d010eeb51/theme/speaker-notes.js#L119-L123

This triggers on any unload of the page - which means prev+next navigation events.

There is a workaround that tries to avoid acting on this, but depending on the the sequence and timing of events this still might break https://github.com/google/comprehensive-rust/blob/c05f0b6f023bf39b8d7f96bfa63e449d010eeb51/theme/speaker-notes.js#L209-L226

**Comment by michael-kerscher** [comment]
To add some thoughts on this: We need some mechanism for the slides to discover that the speaker notes are closed. Currently this is done by signaling back to the slides when the speaker notes close (with the `beforeunload` event - that triggers for any unload, be it window close or some full navigation from the current page).

Options that might be viable and need some discussion:
- quick&dirty: use the current mechanism (see above) that waits for some time but clearly match the events. It currently does not take into account the current state when clearing the timeout (quick navigation can break this with a sequence of "inline-popup" (due to beforeunload event, then speaker notes setup sends "popup" (canceling the timeout) and regular page sends either the "inline-popup" or "inline-closed"  because of this toggling (this seems to trigger a toggle on page load if I see this correctly)https://github.com/google/comprehensive-rust/blob/c05f0b6f023bf39b8d7f96bfa63e449d010eeb51/theme/speaker-notes.js#L90-L93 (this is likely the least effort, a bit brittle depending on timing)
- A ping-pong style polling mechanism based on [BroadcastChannels](https://developer.mozilla.org/en-US/docs/Web/API/BroadcastChannel) where the slides ask regularly (e.g. 500ms) "is there an active speaker notes window" (ping) and the speaker notes answer on that channel with a "yes" (pong). Once that pong is not sent, the state is set to `inline-open` (this seems like the most reliable way and might even recover if a speaker-notes window reappears (think Ctrl+Shift+T -> reopen accidentally closed tab).
- Add some mechanism to remove the `beforeunload` handler when a navigation event happens (more difficult but possible, create a function that is called when click is registered + use that function also when the `current_page` key of the localstorage changes. This could be brittle due to the many ways a navigation event could happen (not only click, but javascript or the back/prev functionality)

**Comment by djmitche** [comment]
I think that `popout` is a [WindowProxy](https://developer.mozilla.org/en-US/docs/Glossary/WindowProxy) object, so it should be possible for the main page to poll that object.

**Comment by michael-kerscher** [comment]
Yes it is, but it is destroyed once the slides is changed as far as I understand. I also thought about using that to figure out if popout.closed is true

**Comment by michael-kerscher** [comment]
This should now be fixed with the latest PR. I wanted to avoid all the described race conditions and used the ping-pong style mechanism described above.
It can even recover itself when the connection was lost. Once a new window/tab with speaker notes is reopened, the broadcast channel is discovering the new window and notes are synced again.
